File: Data Structures & Algorithms/max-area-of-island/submission-10.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        r=len(grid)
        c=len(grid[0])
        def valid(row,col):
            return 0<=row<r and 0<=col<c and grid[row][col]==1
        seen=set()
        directions=[(0,1),(1,0),(-1,0),(0,-1)]
        island=1
        maxisland=0
        def dfs(row,col):
            nonlocal island
            for dx,dy in directions:
                newRow=row+dx
                newCol=col+dy
                if valid(newRow,newCol) and (newRow,newCol) not in seen:
                    seen.add((newRow,newCol))
                    island+=1
                    dfs(newRow,newCol)
        for i in range(r):
            for j in range(c):
                if (i,j) not in seen and valid(i,j):
                    island=1
                    seen.add((i,j))
                    dfs(i,j)
                    maxisland=max(maxisland,island)
        return maxisland

File: Data Structures & Algorithms/rotting-fruit/submission-10.java
class Solution {
    boolean [][] seen;
    public boolean valid(int[][] grid,int row, int col){
        int r=grid.length;
        int c=grid[0].length;
        return (0<=row)&&(row<r)&&(0<=col)&&(col<c)&&(grid[row][col]!=0);

    }
    public int orangesRotting(int[][] grid) {
        int time=0;
        int r=grid.length;
        int c=grid[0].length;
        seen=new boolean[r][c];
        int[][] directions={{0,1},{1,0},{-1,0},{0,-1}};
        Deque<int[]>queue=new ArrayDeque<>();
        for (int i=0;i<r;i++){
            for (int j=0;j<c;j++){
                if (grid[i][j]==2){
                    seen[i][j]=true;
                    queue.offer(new int[]{i,j});
                }
            }
        }
        while (!queue.isEmpty()){
            int length= queue.size();
            boolean rot=false;
            for (int k=0;k<length;k++){
                int[] curr=queue.poll();
                int row=curr[0];
                int col=curr[1];
                for (int[] dir : directions){
                    int newRow=row+dir[0];
                    int newCol=col+dir[1];
                    if (valid(grid,newRow,newCol)&&(seen[newRow][newCol]==false)){
                        rot=true;
                        seen[newRow][newCol]=true;
                        queue.offer(new int[]{newRow,newCol});
                        grid[newRow][newCol]=2;
                    }
                }
            }
            if (rot==true){
                time+=1;
            }
        }
        boolean fresh=false;
        for(int a=0;a<r;a++){
            for (int b=0;b<c;b++){
                if (grid[a][b]==1){
                    fresh=true;
                }
            }
        }
        if (fresh==true){
            return -1;
        }
        return time;

    }
}
